risc0/zkp/verify/riscv.cpp

```diff
@@ -0,0 +1,143 @@
+  MixState assert_zero(Fp4 val, Fp4 mix, const char* loc) const {
+    MixState ret(tot + mul * val, mul * mix);
+#ifdef CIRCUIT_DEBUG
+    if (locs.empty()) {
+      if (val != Fp4()) {
+        ret.locs.emplace_back(loc, val);
+      }
+    } else {
+      ret.locs = locs;
+    }
+#endif
+    return ret;
+  }
+  MixState combine(Fp4 cond, const MixState& inner, const char* loc) const {
+    MixState ret(tot + cond * mul * inner.tot, mul * inner.mul);
+#ifdef CIRCUIT_DEBUG
+    if (locs.empty()) {
+      if (cond != Fp4() && !inner.locs.empty()) {
+        ret.locs = inner.locs;
+        ret.locs.emplace_back(loc, cond);
+      }
+    } else {
+      ret.locs = locs;
+    }
+#endif
+    return ret;
+  }
+};
+
+class RiscVVerifyCircuit : public VerifyCircuit {
+public:
+  RiscVVerifyCircuit(const CodeID& codeID) : codeID_(codeID) {}
+  TapSetRef getTapSet() const override { return getRiscVTaps(); }
+  void execute(ReadIOP& iop) override;
+  void accumulate(ReadIOP& iop) override;
+  Fp4 computePolynomial(const Fp4* evalU, Fp4 polyMix) const override;
+  uint32_t getPo2() const override { return po2_; }
+  bool validCode(const ShaDigest& top) const override;
+
+private:
+  CodeID codeID_;
+  uint32_t po2_;
+  Fp globals_[kGlobalSize];
+};
+
+} // namespace
+
+TapSetRef getRiscVTaps() {
+  static TapSet taps;
+  if (!taps.finalized()) {
+    RegisterGroup accum = RegisterGroup::ACCUM;
+    RegisterGroup code = RegisterGroup::CODE;
+    RegisterGroup data = RegisterGroup::DATA;
+#define tap(base, offset, back) taps.addTap(base, offset, back);
+#define TAPS
+#include "risc0/zkp/prove/step/step.cpp.inc"
+    taps.finalize();
+  }
+  return taps.getRef();
+}
+
+void RiscVVerifyCircuit::execute(ReadIOP& iop) {
+  // Read the low registers + write to globals
+  for (size_t i = 0; i < kOutputRegs; i++) {
+    uint32_t reg;
+    iop.read(&reg, 1);
+    globals_[2 * i] = reg & 0xffff;
+    globals_[2 * i + 1] = reg >> 16;
+  }
+  // Read the po2 size
+  iop.read(&po2_, 1);
+}
+
+void RiscVVerifyCircuit::accumulate(ReadIOP& iop) {
+  // Fill in accum mix
+  for (size_t i = 0; i < kAccumMixGlobalSize; i++) {
+    globals_[kAccumMixGlobalOffset + i] = Fp::random(iop);
+  }
+}
+
+bool RiscVVerifyCircuit::validCode(const ShaDigest& top) const {
+  size_t whichCode = po2_ - log2Ceil(kMinCycles);
+  return codeID_[whichCode] == top;
+}
+
+Fp4 RiscVVerifyCircuit::computePolynomial(const Fp4* evalU, Fp4 polyMix) const {
+  // Do the big polynomial eval
+#define CHECK_EVAL
+#define do_get(buf, reg, back, id) evalU[id];
+#define do_get_global(reg) globals_[reg]
+#define do_begin() MixState()
+#define do_assert_zero(in, val, loc) in.assert_zero(Fp4(val), polyMix, loc)
+#define do_combine(prev, cond, inner, loc) prev.combine(Fp4(cond), inner, loc)
+#define do_add(a, b) Fp4(a) + Fp4(b)
+#define do_sub(a, b) Fp4(a) - Fp4(b)
+#define do_mul(a, b) Fp4(a) * Fp4(b)
+#include "risc0/zkp/prove/step/step.cpp.inc"
+
+#ifdef CIRCUIT_DEBUG
```

Comment:
Curious how you use this with the current bazel build setup. Might be good to add a new `--config=circuit_dbg` mode to `.bazelrc`. (This is a separate PRs worth of work).